src/databricks/labs/ucx/framework/history.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+import dataclasses
+import datetime as dt
+import json
+import logging
+import os
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass
+from functools import cached_property
+from typing import ClassVar, Protocol, TypeVar
+
+from databricks.labs.lsql.backends import SqlBackend
+from databricks.sdk import WorkspaceClient
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class HistoricalRecord:
+    workspace_id: int
+    """The identifier of the workspace where this record was generated."""
+
+    run_id: int
+    """The identifier of the workflow run that generated this record."""
+
+    snapshot_id: int
+    """An identifier that is unique to the records produced for a given snapshot."""
+
+    run_start_time: dt.datetime
+    """When this record was generated."""
+
+    object_type: str
+    """The inventory table for which this record was generated."""
+
+    object_type_version: int
+    """Versioning of inventory table, for forward compatibility."""
+
+    object_id: list[str]
+    """The type-specific identifier for this inventory record."""
+
+    object_data: dict[str, str]
+    """Type-specific data of the inventory record. Keys are top-level attributes, values are their JSON-encoded values."""
+
+    failures: list[str]
+    """The list of problems associated with the object that this inventory record covers."""
+
+    owner: str
+    """The identity of the account that created this inventory record."""
+
+
+class DataclassInstance(Protocol):
+    __dataclass_fields__: ClassVar[dict]
+    # TODO: Once all record types provide the property: key_fields: ClassVar[Sequence[str]]
+
+
+Record = TypeVar("Record", bound=DataclassInstance)
+
+
+class HistoryLog:
+    __slots__ = ("_ws", "_backend", "_run_id", "_catalog", "_schema", "_table")
```

Comment:
This would be the first UCX class that have it, why introduce it here?

Reply:
It's not the first, but I agree it's rare. This is just a way of ensuring that attribute mistakes trigger errors instead of relying on ruff/mypy/pylint to flag them.

Reply:
oke, I guess it is good practice, but also kinda "tech debt" for other classes. Could you create a tech debt issue for this? We might not do this, but let's track it

Reply:
we don't really do this in other places and it doesn't really add to readability. it's not quite consistent with the rest of the codebase.